File: ch2/graph_ds.java
import java.io.*;
import java.util.*;

class pair<X, Y> { // utilizing Java "Generics"
  X _first;
  Y _second;

  public pair(X f, Y s) { _first = f; _second = s; }

  X first() { return _first; }
  Y second() { return _second; }

  void setFirst(X f) { _first = f; }
  void setSecond(Y s) { _second = s; }
}

class graph_ds {
  public static void main(String[] args) throws Exception {
    // Try this input for Adjacency Matrix/List/EL
    // Adj Matrix
    //   for each line: |V| entries, 0 or the weight
    // Adj List
    //   for each line: num neighbors, list of neighbors + weight pairs
    // Edge List
    //   for each line: a-b of edge(a,b) and weight
    /*
    6
      0  10   0   0 100   0
     10   0   7   0   8   0
      0   7   0   9   0   0
      0   0   9   0  20   5
    100   8   0  20   0   0
      0   0   0   5   0   0
    6
    2 2 10 5 100
    3 1 10 3 7 5 8
    2 2 7 4 9
    3 3 9 5 20 6 5
    3 1 100 2 8 4 20
    1 4 5
    7
    1 2 10
    1 5 100
    2 3 7
    2 5 8
    3 4 9
    4 5 20
    4 6 5
    */

    File f = new File("graph_ds.txt");
    Scanner sc = new Scanner(f);
    int V = sc.nextInt(); // we must know this size first!
                          // remember that if V is > 100, try NOT to use AM!
    int[][] AM = new int[V][];
    for (int i = 0; i < V; i++) {
      AM[i] = new int[V];
      for (int j = 0; j < V; j++)
        AM[i][j] = sc.nextInt();
    }

    System.out.println("Neighbors of vertex 0:");
    for (int j = 0; j < V; j++) // O(|V|)
      if (AM[0][j] != 0)
        System.out.println("Edge 0-" + j + " (weight = " + AM[0][j] + ")");

    V = sc.nextInt();
    Vector< Vector< pair < Integer, Integer > > > AL = new Vector< Vector< pair < Integer, Integer > > >(V);
    for (int i = 0; i < V; i++) { // for each vertex
      Vector< pair < Integer, Integer > > Neighbor = new Vector < pair < Integer, Integer > >();
      AL.add(Neighbor); // add this empty neighbor list to Adjacency List
    }

    for (int i = 0; i < V; i++) {
      int total_neighbors = sc.nextInt();
      for (int j = 0; j < total_neighbors; j++) {
        int id = sc.nextInt();
        int weight = sc.nextInt();
        AL.get(i).add(new pair<Integer, Integer>(id-1, weight)); // some index adjustment
      }
    }

    System.out.println("Neighbors of vertex 0:");
    for (pair<Integer, Integer> v_w : AL.get(0)) {
      int v = v_w.first(), w = v_w.second();
      System.out.println("Edge 0-" + v + " (weight = " + w + ")");
    }

    int E = sc.nextInt();
    PriorityQueue<pair<Integer, pair<Integer, Integer>>> EL = new PriorityQueue<>(1, 
      new Comparator<pair<Integer, pair<Integer, Integer>>>() { // overriding the compare method
        public int compare(pair<Integer, pair<Integer, Integer>> i, pair<Integer, pair<Integer, Integer>> j) {
          return i.first() - j.first(); // currently min heap based on cost
        }
      }
    );

    for (int i = 0; i < E; i++) {
      int a = sc.nextInt();
      int b = sc.nextInt();
      pair<Integer, Integer> ab = new pair<>(a, b);
      int weight = sc.nextInt();
      EL.offer(new pair<Integer, pair<Integer, Integer>> (-weight, ab) ); // to reverse sort order
    }

    // edges sorted by weight (smallest->largest)
    for (int i = 0; i < E; i++) {
      pair < Integer, pair < Integer, Integer > > edge = EL.poll();
      // negate the weight again
      System.out.println("weight: " + (-edge.first()) + " (" + edge.second().first() + "-" + edge.second().second() + ")");
    }
  }
}

File: ch2/ownlibrary/graph_ds.ml
open Printf
open Scanf

module IIIMinPQ = Set.Make (struct
    type t = int * int * int
    let compare = compare
  end)

let () =
 (* Try this input for Adjacency Matrix/Adjacency List/Edge List
  * Adjacency Matrix am
  *   for each line: |v| entries, 0 or the weight
  * Adjacency List al
  *   for each line: num neighbors, list of neighbors + weight pairs
  * Edge List el
  *   for each line: a-b of edge(a,b) and weight *)

  (* 6
   *   0  10   0   0 100   0
   *  10   0   7   0   8   0
   *   0   7   0   9   0   0
   *   0   0   9   0  20   5
   * 100   8   0  20   0   0
   *   0   0   0   5   0   0
   * 6
   * 2 2 10 5 100
   * 3 1 10 3 7 5 8
   * 2 2 7 4 9
   * 3 3 9 5 20 6 5
   * 3 1 100 2 8 4 20
   * 1 4 5
   * 7
   * 1 2 10
   * 1 5 100
   * 2 3 7
   * 2 5 8
   * 3 4 9
   * 4 5 20
   * 4 6 5 *)

  let ic = Scanning.open_in "graph_ds.txt" in

  let v = bscanf ic "%d\n" (fun x -> x) in  (* we must know this size first! *)
                         (* remember that if v is > 2000, try NOT to use am! *)
  let am = Array.init v (fun _ ->
      Array.init v (fun _ ->
          bscanf ic " %d" (fun x -> x)
        )
    ) in

  printf "Neighbors of vertex 0:\n";
  am.(0) |> Array.iteri (fun j x ->  (* O(|v|) *)
      if x <> 0 then
        printf "Edge 0-%d (weight = %d)\n" j x
    );

  let v = bscanf ic " %d\n" (fun x -> x) in
  let al = Array.init v (fun _ ->
    let total_neighbors = bscanf ic "%d " (fun x -> x) in
    List.init total_neighbors (fun _ ->
        bscanf ic "%d %d " (fun id weight -> (id - 1, weight))  (* some index adjustment *)
      )
    ) in
  printf "Neighbors of vertex 0:\n";
  al.(0) |> List.iter (fun (v, w) ->
      (* al.(0) contains the required information *)
      (* O(k), where k is the number of neighbors *)
      printf "Edge 0-%d (weight = %d)\n" v w
    );

  let e = bscanf ic "%d\n" (fun x -> x) in
  let el = List.init e (fun _ ->
      bscanf ic "%d %d %d\n" (fun a b weight -> (weight, a, b))
    ) |> IIIMinPQ.of_list  (* one way to store Edge list *)
  in

  (* edges sorted by weight (smallest->largest) *)
  el |> IIIMinPQ.iter (fun (weight, a, b) ->
      printf "weight: %d (%d-%d)\n" weight a b
    );
